docs/orchestrating-elastic-stack-applications/elasticsearch/orchestration.asciidoc

```diff
@@ -173,3 +173,68 @@ Specifically, if the Elasticsearch index settings demand a higher number of shar
 
 ** Adjust the Elasticsearch link:https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-update-settings.html[index settings] to a number of replicas that allow the desired node removal.
 ** Use link:https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules.html#dynamic-index-settings[`auto_expand_replicas`] to automatically adjust the replicas to the number of data nodes in the cluster.
+
+[id="{p}-advanced-upgrade-control"]
+== Advanced control during rolling upgrades
+
+The rules (otherwise known as predicates) that the ECK operator follows during an Elasticsearch upgrade can be selectively disabled for certain scenarios where the ECK operator will not proceed with an Elasticsearch cluster upgrade because it deems it to be "unsafe".
+
+WARNING: Selectively disabling the predicates below are extremely risky, and carry a high chance of either data loss, or causing a cluster to become completely unavailable.  Use only if you are absolutely positive you are not causing permanent damage to an Elasticsearch cluster.
+
+* The following named predicates control the upgrade process
+
+** do_not_restart_healthy_node_if_MaxUnavailable_reached
++
+If `maxUnavailable` is reached, only allow unhealthy Pods to be deleted.
+** skip_already_terminating_pods
++
+Do not attempt to restart pods that are already in the process of being terminated.
+** only_restart_healthy_node_if_green_or_yellow
++
+Only restart healthy Elasticsearch nodes if the health of the cluster is either green or yellow, never red.
+** if_yellow_only_restart_upgrading_nodes_with_unassigned_replicas
++
+During a rolling upgrade, primary shards assigned to a node running a new version cannot have their replicas assigned to a node with the old version. Therefore we must allow some Pods to be restarted even if cluster health is Yellow so the replicas can be assigned.
+** require_started_replica
++
+If a cluster is yellow, allow deleting a node, but only if they do not contain the only replica of a shard since it would make the cluster go red.
+** one_master_at_a_time
++
+Only allow a single master to be upgraded at a time.
+** do_not_delete_last_master_if_all_master_ineligible_nodes_are_not_upgraded
++
+Force an upgrade of all the master-ineligible nodes before upgrading the last master-eligible node.
+** do_not_delete_pods_with_same_shards
```

Comment:
As the predicate names become public with this feature, I wonder if we should write a little check to run in CI to verify that the names in the doc are in sync with the names in the code.

Something like this:
```bash
code=pkg/controller/elasticsearch/driver/upgrade_predicates.go \
doc=docs/orchestrating-elastic-stack-applications/elasticsearch/orchestration.asciidoc \
predicate_pattern='[a-z]*_[A-Za-z_]*'
diff \
	<(grep "name:" "$code" | grep -o "$predicate_pattern" ) \
	<(grep '\*\* [a-z]' "$doc" | grep -o "$predicate_pattern" )
```
If we add a comment in the two files to mark the predicate name (`// predicate`), this could make the parsing more sure. 
```
diff \
	<(grep '// predicate' $code | grep -o "$predicate_pattern" ) \
	<(grep '// predicate' -A1 $doc | grep -o "$predicate_pattern" )
```

Reply:
This is a good thought. I've added this CI check, missing the marks.  Let's see if that's sufficient.  I'm happy to also add the marks if we feel that's necessary.